recipe/2023-11-validate-template-path.md

```diff
@@ -0,0 +1,96 @@
+
+#### User story 2
+
+As a Radius user, I want to define a Terraform recipe with a Terraform module template path. My module source is not the supported Terraform registry or HTTP URLs (i.e. Github, local paths, S3 buckets, etc). When I try to register the recipe to my environment, I recieve an error message notifying me that I can only provide a Terraform module registry or HTTP URL as a valid module source using Radius. 
+
+## Design
+
+### Design details
+
+### API design (if applicable)
+N/A. We are updating an existing API to include more validation, but this won't include any public-facing or user-facing changes. 
+
+
+#### Validate module sources as part of the CreateOrUpdateEnvironment API endpoint 
+Right now, we only return an error at deployment time. We'd like to return an error while the recipe is being registered to the environment. To do so, we'll add validation as part of the `CreateOrUpdateEnvironment` API. This requires no new API changes, just an update to how we handle environment conversion.
+
+We already have a check for local path module sources [here](https://github.com/radius-project/radius/blob/40c91fdc3a4dd3ac04906094dc8302f7232d700d/pkg/corerp/api/v20231001preview/environment_conversion.go#L303). Instead of just rejecting local module source paths, we'll only accept Terraform registries and HTTP URLs and reject everything else.  
+
+#### Validation steps
+Terraform registries: 
+Terraform has code to check whether a module source is a Terraform recipe as part of the `terraform-registry-address` [package](https://github.com/hashicorp/terraform-registry-address/blob/main/module.go#L44). We will use this package to validate Terraform module registries. 
+
+HTTP URLs: 
+Terraform treats HTTP URLs as remote addresses and doesn't have functionality for checking HTTP URLs in isolation. We will need to implement our own validation for HTTP URLs. Go provides validation for URIs as part of the `url` [package](https://pkg.go.dev/net/url#ParseRequestURI), but we will likely need additional checks to ensure that a `.zip`/`.tar`/etc suffix is present as well. 
+
+Any module source that is not a Terraform registry or an HTTP URL will be treated as unsupported. We will return a `NewClientErrInvalidRequest` [error](https://github.com/radius-project/radius/blob/main/pkg/armrpc/api/v1/error.go#L64) in this case. The error message will indicate that we only support Terraform registries and HTTP URLs currently. 
+
+## Alternatives considered
+
+1. Adding a `ValidateTemplateKind` API endpoint (similar to `GetRecipeMetadata`).
+This was rejected because the logic could be added to the existing `CreateOrUpdateEnvironment` API endpoint instead. That way, there's less code to add and we can return a validation error before we save any environment information. 
+2. Support all Terraform module sources instead of restricting supported sources.
```

Comment:
A big reason why it seems ok to limit this is that we haven't done the work to enable private sources yet. Github, http(s) and public registry are the main sources for public modules. 

We'll need to revisit the set of supported sources in the future when we enable private sources.